packages/compass-components/src/components/bson-document-editor.tsx

```diff
@@ -0,0 +1,180 @@
+import React, { useRef, useState } from 'react';
+import { css } from '@leafygreen-ui/emotion';
+import { spacing } from '@leafygreen-ui/tokens';
+import type { EditorProps } from './editor';
+import { Editor } from './editor';
+import queryParser from 'mongodb-query-parser';
+import { EJSON } from 'bson';
+import { useElementParentHoverState } from '../utils/use-element-parent-hover-state';
+import { Button } from './leafygreen';
+
+type BSONValue = any;
+
+type BSONDocumentEditorProps = Omit<EditorProps, 'variant'> & {
+  variant?: never;
+  onChangeValue?: (
+    value: BSONValue,
+    error: Error | null,
+    variant: 'EJSON' | 'Shell',
+    text: string,
+    event?: any
+  ) => void;
+} & (
+    | {
+        initialValue?: never;
+      }
+    | {
+        text?: never;
+        initialValue?: BSONValue;
+      }
+  );
+
+const actionsGroupContainer = css({
+  position: 'absolute',
+  display: 'flex',
+  gap: spacing[2],
+  width: '100%',
+  top: spacing[3],
+  right: spacing[3],
+  paddingLeft: spacing[3],
+  paddingRight: spacing[3],
+  pointerEvents: 'none',
+  zIndex: 100,
+});
+
+const actionsGroupItem = css({
+  flex: 'none',
+  pointerEvents: 'all',
+});
+
+function BSONDocumentEditor({
+  initialValue,
+  onChangeValue,
+  text,
+  onChangeText,
+  ...otherProps
+}: BSONDocumentEditorProps): React.ReactElement {
+  const containerRef = useRef<HTMLDivElement | null>(null);
+  const isHovered = useElementParentHoverState(containerRef);
+
+  if (!text && initialValue) {
+    text = queryParser.toJSString(initialValue);
+  }
+
+  const [currentVariant, setCurrentVariant] = useState<'EJSON' | 'Shell'>(
+    'Shell'
+  );
+  const [lastAlternateVariantRepr, setLastAlternateVariantRepr] = useState<{
+    text: string;
+    value: BSONValue;
+  } | null>(null);
+  const [currentText, setCurrentText] = useState(text ?? '');
```

Comment:
Never quite sure about whether I’m doing the right things state-management-wise. :)

Reply:
I think this looks alright, maybe one thing worth noting is that the way it's currently implemented this `text` property is more like `initialText` (or `defaultText` if we want to follow React controlled / uncontrolled components patterns)